src/Raven.Server/Documents/Indexes/MapReduce/AutoMapReduceIndex.cs

```diff
@@ -230,88 +208,49 @@ private Table GetMapEntriesTable(Transaction tx)
-
-            table.Insert(tvb);
-            
-            var pos = state.Tree.DirectAdd(new Slice((byte*) &etag, sizeof (long)), mappedResult.Size);
-
-            mappedResult.CopyTo(pos);
-        }
 
-        public unsafe List<MapEntry> GetMapEntriesForDocument(Table table, LazyStringValue documentKey)
-        {
-            var result = new List<MapEntry>();
-
-            var documentKeySlice = new Slice(documentKey.Buffer, (ushort) documentKey.Size);
+            var existingEntry = GetMapEntryForDocument(mapEntriesTable, documentKey);
 
-            var seekForwardFrom = table.SeekForwardFrom(_mapResultsSchema.Indexes["DocumentKeys"], documentKeySlice);
-
-            foreach (var seek in seekForwardFrom)
+            long storageId;
+            if (existingEntry == null)
+                storageId = mapEntriesTable.Insert(tvb);
+            else
             {
-                if (seek.Key.Equals(documentKeySlice) == false)
-                    break;
-
-                foreach (var tvr in seek.Results)
-                {
-                    int _;
-                    var ptr = tvr.Read(0, out _);
-                    var etag = IPAddress.NetworkToHostOrder(*(long*)ptr);
-
-                    ptr = tvr.Read(2, out _);
-                    var reduceKeyHash = Bits.SwapBytes(*(ulong*) ptr);
-
-                    result.Add(new MapEntry
-                    {
-                        Etag = etag,
-                        ReduceKeyHash = reduceKeyHash,
-                        StorageId = tvr.Id
-                    });
-                }
+                // no need to update since we have the same entry already stored
+                storageId = existingEntry.StorageId;
             }
 
-            return result;
+            var pos = state.Tree.DirectAdd(new Slice((byte*)&storageId, sizeof(long)), mappedResult.Size);
+
+            mappedResult.CopyTo(pos);
         }
 
-        protected override unsafe void LoadValues()
+        public unsafe MapEntry GetMapEntryForDocument(Table table, LazyStringValue documentKey)
         {
-            base.LoadValues();
+            var documentKeySlice = new Slice(documentKey.Buffer, (ushort) documentKey.Size);
 
-            TransactionOperationContext context;
-            using (_contextPool.AllocateOperationContext(out context))
-            using (var tx = context.OpenReadTransaction())
-            {
-                var tree = tx.InnerTransaction.ReadTree("MapResults");
+            var seek = table.SeekForwardFrom(_mapResultsSchema.Indexes["DocumentKeys"], documentKeySlice).FirstOrDefault();
 
-                if (tree == null)
-                    return;
+            if (seek?.Key.Compare(documentKeySlice) != 0)
+                return null;
 
-                var table = GetMapEntriesTable(tx.InnerTransaction);
-                
-                if (table.NumberOfEntries == 0)
-                    return;
+            var tvr = seek.Results.Single();
```

Comment:
No, doing aggregation during the map is good, it can help us somewhat.
But remember that with real indexes, we are going to be having people that do fanout from the reduce.